### DISCUSSION TITLE
error: ‘EIGEN_DEVICE_FUNC’ does not name a type

### DISCUSSION THREAD

**Discussion by dangla** [autodiff/autodiff#325] [Q&A]
Hi
autodiff seems a very useful tools. However during the installation I got the message above which comes out from the header Macros.h from eigen3. I don't have any idea how to solve this. Could you help me?
Patrick

**Comment by allanleal** [comment]
Please describe the step by step that caused this.

**Comment by dangla** [comment]
I first loaded some missing libraries 
- eigen3 (libeigen3-dev version 3.3.72)
- pybind11-dev version 2.4.3-2build2
Then installed autodiff with cmake (on linux) following the procedure i.e.:
mkdir .build
cd .build and make ..
cmake --build . --target install
This last command gives the error message.



- autodiff with cmake.

**Reply by allanleal** [reply]
Please download the latest source code. I just pushed a commit that should fix this issue for you. Let me know.

**Reply by dangla** [reply]
OK. Thanks.

**Reply by dangla** [reply]
I just tested it. But the error still remains. I get the compiling error:

In file included from /home/dangla/Documents/Softwares/bil/DualNumbers/autodiff-main/autodiff/common/meta.hpp:42,
                 from /home/dangla/Documents/Softwares/bil/DualNumbers/autodiff-main/tests/common/meta.test.cpp:34:
/usr/include/eigen3/Eigen/src/Core/util/Macros.h:615:26: error: ‘EIGEN_DEVICE_FUNC’ does not name a type
  615 |     template<typename T> EIGEN_DEVICE_FUNC void ignore_unused_variable(const T&) {}

It seems that the macro EIGEN_DEVICE_FUNC is not known at the entry of the header /usr/include/eigen3/Eigen/src/Core/util/Macros.h.

Patrick

**Comment by allanleal** [comment]
Did you download the latest commit or the latest release? You need to use
this link (or do git clone):
https://github.com/autodiff/autodiff/archive/refs/heads/main.zip

**Reply by dangla** [reply]
Yes I download the last commit (not the release!). I get the same error.
Error is in meta.test.cpp!
:((

**Comment by allanleal** [comment]
Ignore this error for now. That's a test file. You don't need it in your
application. I'll check later what may be happening. Compilation here is
successful.

**Reply by dangla** [reply]
I saw that EIGEN_DEVICE_FUNC is defined in <Eigen/Core> not in <Eigen/src/Core/util/Macros.h>. 
Maybe it can help!

**Reply by dangla** [reply]
I have C++14. Should it run with it?

**Reply by dangla** [reply]
I tried in my application. It works provided that I define the macro EIGEN_DEVICE_FUNC before calling the header dual.hpp, i.e.

#define  EIGEN_DEVICE_FUNC
#include <autodiff/forward/dual.hpp>
....

because EIGEN_DEVICE_FUNC is not defined at the entry of <Eigen/src/Core/util/Macros.h> which is called in meta.hpp.
I might do something wrong!!

**Comment by allanleal** [comment]
Please check the `meta.hpp` file in your computer. See if the following is present:

~~~cpp
#ifdef EIGEN_DEVICE_FUNC
#define AUTODIFF_DEVICE_FUNC EIGEN_DEVICE_FUNC
#else
#define AUTODIFF_DEVICE_FUNC
#endif
~~~

This was a change I introduced yesterday. It should have resolved your issue.

**Reply by dangla** [reply]
It is present but the issue comes in <Eigen/src/Core/util/Macros.h> and in this header the macro EIGEN_DEVICE_FUNC must be defined.

**Reply by dangla** [reply]
Just FYI
by calling <autodiff/forward/dual/eigen.hpp> first and then <autodiff/forward/dual.hpp> it works (the macro EIGEN_DEVICE_FUNC is defined by passing through <autodiff/forward/dual/eigen.hpp>).
Thanks a lot for helping.
